Add caching for images

furbrain opened the pull request:
This pull request implements an ImageCache - this can retrieve files from the web or local storage, but will not hold more than approx 2Mb in memory at any one time and will drop the oldest images if this limit will be over-reached. Images retrieved from the web will be "re-retrieved" if they have exceeded their cache times as per http protocols. Images from local storage will be reloaded if they have changed on-disk,

This PR addresses the second part of issue #13  - intelligent caching

furbrain committed:
Add caching - first iteration

furbrain committed:
Add tests for cache.py

furbrain committed:
Ignore egg files

furbrain committed:
Merge branch 'master' of https://github.com/tingbot/tingbot-python into temp_master

furbrain committed:
ignore .eggs

furbrain committed:
Add caching size limit and integrate with Screen

furbrain committed:
bug fixes

furbrain committed:
remove _is_url: no longer needed

joerick commented:
Taking a look at this now 😎

joerick commented on tingbot/cache.py:
```diff
@@ -0,0 +1,135 @@
+        return self.image  
+        
+    def get_size(self):
+        return self.image.surface.get_buffer().length  
```
This will only count 1 frame for GIFs - since the `surface` variable changes depending on the time. I think this would be better as a property on Image and GIFImage classes

furbrain replied:
Agreed

joerick commented on tingbot/cache.py:
```diff
@@ -0,0 +1,135 @@
+            return False
+        return False
+        
+class FileImage(object):
```
Should inherit from `ImageEntry`?

furbrain replied:
Yup, typo

joerick commented on tingbot/cache.py:
```diff
@@ -0,0 +1,135 @@
+        self.last_accessed=time.time()
+
+    def is_fresh(self):
+        return self.last_modified==os.path.getmtime(self.filename)       
```
This could raise an error - `OSError: [Errno 2] No such file or directory` - if the image has moved/deleted. Should be okay to return False in that case, I think.

joerick commented on tingbot/cache.py:
```diff
@@ -0,0 +1,135 @@
+            if self.images[location].is_fresh():
+                return self.images[location].get_image()
+            else:
+                sel.del_image(location)
```
Typo?

furbrain replied:
Yup

joerick commented on tingbot/cache.py:
```diff
@@ -0,0 +1,135 @@
+from urlparse import urlparse
+
+def get_http_timestamp(dt):
+    return calendar.timegm(rfc822.parsedate(dt))
```
A few time comparisons in this file between different timebases (local vs. server) - could use the Date header in the HTTP response to correct the offset before storing.

furbrain replied:
Good point. I'll need to look at this tomorrow. Have fixed the other errors and will commit relevant changes

joerick replied:
:+1:

furbrain replied:
I have now addressed this issue. All time comparisons should now be done in the relevant (local or server) timebase.

furbrain committed:
Bug fixes

furbrain committed:
Bug fix

furbrain committed:
Add get_server_date - gets server time of a request

This allows the rest of the code to compare server-provided times with an
appropriate "now" for that server

furbrain committed:
bug fix

joerick commented on tingbot/cache.py:
```diff
@@ -0,0 +1,145 @@
+        self.retrieved = time.time() #local time
+        
+    def set_attributes(self,response):    
+        self.last_modified = get_last_modified(response) #server time
```
Could `last_modified` be stored in the local timebase too? It would be good if the timebase stuff was handled by the accessors `get_last_modified` etc, so we don't have to think the different timebases in this class

furbrain replied:
ok, will do

furbrain replied:
Actually, looking at this, we are mostly using last_modified to compare to different request responses. These may have different offsets from our local timebase (because we are comparing with when it was received - which is not necessarily when sent and is dependent on network factors). This means that out "local" last_modified may change from request to request, even though the last_modified on the server has remained the same.

For this reason, I think we should keep this as a "server time" variable (and keep it documented as such in the code)

joerick replied:
Yeah, agreed. My bad.

joerick committed:
Whitespace

joerick committed:
Mock is also required for the tests

joerick committed:
Remove unused import

joerick commented on tingbot/graphics.py:
```diff
@@ -410,7 +395,7 @@ def _get_frames(self, pil_image):
                 palette = base_palette
 
-            pygame_image = pygame.image.fromstring(pil_image.tostring(), pil_image.size, pil_image.mode)
+            pygame_image = pygame.image.fromstring(pil_image.tobytes(), pil_image.size, pil_image.mode)
```
What's the thinking behind this change? I'm getting `AttributeError: tobytes` on my machine.

furbrain replied:
The thinking was that we should prefer using a `bytes` data type rather than `str` in anticipation of transferring to python3. However, the tobytes function is only available in pillow (not PIL) which is what is actually installed on my machine. So I should probably revert this change for now.

joerick replied:
Cool, makes sense. I've got Pillow 1.7.8 on here. Guess we should put version ranges in the setup.py when we do the switch to Python 3. Let's revert this for now, we can address it later.

furbrain replied:
So I tried to revert it back... and got 

```
Exception: tostring() has been removed. Please call tobytes() instead.
```

Sigh. I'll put in  a try:except: block to cope with both situations. I've added a test so that any problems get picked up. 
I'm on Pillow 3.1.2

joerick commented:
I wanted to know how the cache would cope with a really large image.

```
screen.image('gif_large.gif', scale=0.5)
```

([The large GIF](https://upload.wikimedia.org/wikipedia/commons/0/0b/New_cyclic_large.gif) I used for this - 12MB on disk, 42MB in memory as reported by `get_memory_usage`)

I got the following error:

```
Traceback (most recent call last):
  File "/Users/joerick/Work/Tingbot/6 tingbot-lib-dev/tingbot/run_loop.py", line 65, in run
    next_timer.action()
  File "gif_large_test.py", line 6, in loop
    screen.image('gif_large.gif', scale=0.5)
  File "/Users/joerick/Work/Tingbot/6 tingbot-lib-dev/tingbot/graphics.py", line 270, in image
    super(Screen, self).image(*args, **kwargs)
  File "/Users/joerick/Work/Tingbot/6 tingbot-lib-dev/tingbot/graphics.py", line 212, in image
    image = image_cache.get_image(image)
  File "/Users/joerick/Work/Tingbot/6 tingbot-lib-dev/tingbot/cache.py", line 143, in get_image
    return self.images[location].get_image()
KeyError: 'gif_large.gif'
```

I think it was deleted out of the cache before the function returned.

joerick commented:
I also did a little bit of cleanup as I was testing - pull from branch [pull-26](https://github.com/tingbot/tingbot-python/tree/pull-26) before making more changes please. Thanks!

furbrain committed:
Merge pull request #3 from tingbot/pull-26

Merge code tidies from @joerick

furbrain committed:
Fix bug due to different Pillow APIs

furbrain committed:
Fix bug when loading an image larger than the cache max_size

furbrain commented:
Ok, have pulled in data from pull-26, fixed bug with GIF images and with large files and added tests for this

furbrain commented:
One question - do we need to make the ImageCache thread safe? I'm trying to think of a situation where it would break...

furbrain commented:
Got it - two threads calling get_image and exceeding the size of the cache, triggers two competing loops clearing out the cache. I'll lock the get_image and del_image functions

furbrain committed:
Add thread safety for ImageCache

joerick commented:
Merged 👍 cheers @furbrain!
